controllers/machinedeployment_controller.go

```diff
@@ -213,6 +236,80 @@ func (r *MachineDeploymentReconciler) reconcile(ctx context.Context, cluster *cl
 	return ctrl.Result{}, errors.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
+func (r *MachineDeploymentReconciler) rolloutAfter(ctx context.Context, machineDeployment *clusterv1.MachineDeployment, newMachineSet *clusterv1.MachineSet) error {
+	var machinesToRollout []*clusterv1.Machine
+	machines, err := r.getMachinesForMachineSet(ctx, newMachineSet)
+	if err != nil {
+		return err
+	}
+
+	// Find Machines needing to rollout after.
+	for key := range machines {
+		if machines[key].DeletionTimestamp.IsZero() &&
+			machines[key].CreationTimestamp.Before(machineDeployment.Spec.RolloutAfter) {
+			machines[key].Annotations[clusterv1.DeleteMachineAnnotation] = "True"
+			// TODO: Patch machines
+			machinesToRollout = append(machinesToRollout, machines[key])
+		}
+	}
+
+	// Scale up if we can
+	maxSurge, err := intstrutil.GetScaledValueFromIntOrPercent(machineDeployment.Spec.Strategy.RollingUpdate.MaxSurge, int(*(machineDeployment.Spec.Replicas)), true)
+	if err != nil {
+		return err
+	}
+	maxTotalMachines := *(machineDeployment.Spec.Replicas) + int32(maxSurge)
```

Comment:
I'm wondering if the MachineSet will have the right size after the up and downscale considering the potential states that can occur during this procedure. Is there some kind of fail safe that the MachinSet will eventually converge onto the correct replica count again?